src/wasm-interpreter.h

```diff
@@ -2680,899 +2681,875 @@ template<typename GlobalManager, typename SubType> class ModuleInstanceBase {
         }
       }
     }
-  };
 
-  // Executes expressions with concrete runtime info, the function and module at
-  // runtime
-  template<typename RERSubType>
-  class RuntimeExpressionRunnerBase : public ExpressionRunner<RERSubType> {
-    ModuleInstanceBase& instance;
-    FunctionScope& scope;
-    // Stack of <caught exception, caught catch's try label>
-    SmallVector<std::pair<WasmException, Name>, 4> exceptionStack;
+    ~FunctionScope() { parent.scope = oldScope; }
+
     // The current delegate target, if delegation of an exception is in
     // progress. If no delegation is in progress, this will be an empty Name.
+    // This is on a function scope because it cannot "escape" to the outside,
+    // that is, a delegate target is like a branch target, it operates within
+    // a function.
     Name currDelegateTarget;
+  };
 
-  protected:
-    // Returns the instance that defines the memory used by this one.
-    SubType* getMemoryInstance() {
-      auto* inst = instance.self();
-      while (inst->wasm.memory.imported()) {
-        inst = inst->linkedInstances.at(inst->wasm.memory.module).get();
-      }
-      return inst;
-    }
+private:
+  // This is managed in an RAII manner by the FunctionScope class.
+  FunctionScope* scope = nullptr;
 
-    // Returns a reference to the current value of a potentially imported global
-    Literals& getGlobal(Name name) {
-      auto* inst = instance.self();
-      auto* global = inst->wasm.getGlobal(name);
-      while (global->imported()) {
-        inst = inst->linkedInstances.at(global->module).get();
-        Export* globalExport = inst->wasm.getExport(global->base);
-        global = inst->wasm.getGlobal(globalExport->value);
-      }
+  // Stack of <caught exception, caught catch's try label>
+  SmallVector<std::pair<WasmException, Name>, 4> exceptionStack;
+
+protected:
+  // Returns the instance that defines the memory used by this one.
+  SubType* getMemoryInstance() {
+    auto* inst = self();
```

Comment:
Probably no need for a separate `inst` variable here and below.

Reply:
`inst` is modified (`inst = ..`) in the loop. We are basically starting at this instance, and then looking through the imports to find the proper instance that defines the memory (which may be this one, or another).

Reply:
I think this is the only open question on this PR - any remaining concerns after my response?

Reply:
ah, thanks. Makes sense.